File: 2553-separate-the-digits-in-an-array/solution.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<int> separateDigits(std::vector<int>& nums) {
        std::vector<int> result;
        
        // Iterate through each number in the input array
        for (int num : nums) {
            // Temporary storage for digits of the current number
            // because mathematical extraction gives digits in reverse order
            std::vector<int> currentDigits;
            
            while (num > 0) {
                currentDigits.push_back(num % 10);
                num /= 10;
            }
            
            // Add digits to result in the correct order (left-to-right)
            for (int i = currentDigits.size() - 1; i >= 0; --i) {
                result.push_back(currentDigits[i]);
            }
        }
        
        return result;
    }
};

File: 2553-separate-the-digits-in-an-array/solution.py
class Solution:
    def separateDigits(self, nums: list[int]) -> list[int]:
        """
        Approach: Convert each integer to a string to easily iterate over its digits,
        then convert each character back to an integer and flatten the list.
        """
        answer = []
        
        for num in nums:
            # Convert number to string to access individual digits
            s_num = str(num)
            for char in s_num:
                # Convert back to integer and append to result
                answer.append(int(char))
                
        return answer

    def separateDigitsAlternative(self, nums: list[int]) -> list[int]:
        """
        Alternative: One-liner using list comprehension for Pythonic efficiency.
        """
        return [int(digit) for num in nums for digit in str(num)]